README.md
[![Build Status](https://travis-ci.org/esa-esdl/xcube-viewer.svg?branch=master)](https://travis-ci.org/esa-esdl/xcube-viewer)

# ESDL xcube-viewer

A simple viewer component for Xcube server.

![xcube-viewer](./doc/xcube-viewer.jpg)

## Run it

1. Install [xcube](https://github.com/dcs4cop/xcube) and run server in demo mode
2. Install and start `xcube-viewer` with demo configuration


### `xcube`

`xcube` one-time install:

    $ git clone https://github.com/dcs4cop/xcube.git
    $ cd xcube
    $ conda env create
    $ conda activate xcube
    $ python setup.py develop  

    
`xcube` update and run server:

    $ cd xcube
    $ git pull
    $ conda activate xcube  
    $ xcube serve --verbose --traceperf --config xcube/webapi/res/demo/config.yml  
    
If errors occur, you may need to update the environment:

    $ conda env update

### `xcube-viewer`

Checkout `xcube-viewer` sources:

    $ git clone https://github.com/dcs4cop/xcube-viewer.git
    $ cd xcube-viewer
    $ yarn install
    $ yarn start


Generate file `.env.local` with following content:

    REACT_APP_TITLE=xcube Viewer (dev)

The following entries are optional, but required when enabling user management through [auth0](https://auth0.com/):

    REACT_APP_AUTH0_DOMAIN=<YOUR AUTH0 DOMAIN>
    REACT_APP_AUTH0_CLIENT_ID=<YOUR AUTH0 CLIENT ID>
    REACT_APP_AUTH0_AUDIENCE=<YOUR AUTH0 API ID>

Update, install, and run:

    $ cd xcube-viewer
    $ git pull
    $ yarn install
    $ yarn start


Build `xcube-viewer` for deployment with `default` branding:

    $ cd xcube-viewer
    $ git pull
    $ yarn build
    
Find outputs in `./build`. To build different branding targets, create a file `.env.local` and specify
the branding, e.g.

    REACT_APP_BRANDING=eodatabee
    
will build target `eodatabee`. Available targets are `default`, `cyanoalert`, `eodatabee`, `esdl`.

To add a new branding, choose a branding name and create directories with that name in `public` 
and in `src/resources`. It is easiest to copy the `default` branding and rename it. 
Then alter what need be branded, e.g. typically the logos.

You will also need to add a new entry in object named `brandings` in `src/config.ts` 
using your branding name as key. Also here, it is best to copy the `default` one and adapt it
to your needs.       

## Features

* Select dataset, update variable and places list, fly to bounding box
* Select place (Lakes / Stations) and fly to bounding box
* Select variable and show as layer on map
* Select time and update variable layer on map
* Click in map and show time series chart
* Select a place and show time series chart
* Select dataset and fly to bounding box
* Select place and fly to bounding box
* Click in time series chart and select time
* Show selected time in time series chart
* Zoom into and out of time series chart
* Use local time in UI (backend provides UTC time only)
* Show a variable's legend in map
* Switch language. Texts are internationalizable. Locales are currently "en" and "de". (see constant `I18N` in `src/config/config.ts`)
* Show modal "loading" dialog on initial data load
* Show toast on info events (e.g. server error) 
* Animate through time (play forw/back)
* Added means to select a time slice
* Time ranges set from datasets
* Change xcube server. Users' xcube server configurations are saved in local browser storage.  
* Place groups associated to datasets
  - produce a unique vector layer in the map;
  - can specify their feature property fields that will provide the values for predefined roles:
    "label", "infoUrl" (not used yet);
* User management

## More

* [User Guide](https://xcube.readthedocs.io/en/latest/viewer.html#)
* [Planned Enhancements](https://github.com/dcs4cop/xcube-viewer/labels/enhancement)
* [Known Issues](https://github.com/dcs4cop/xcube-viewer/labels/bug)


--- 



This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app) v3.3.0.

## Available Scripts

In the project directory, you can run:

### `yarn start`

Runs the app in the development mode.<br />
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.<br />
You will also see any lint errors in the console.

### `yarn test`

Launches the test runner in the interactive watch mode.<br />
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `yarn build`

Builds the app for production to the `build` folder.<br />
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.<br />
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `yarn eject`

**Note: this is a one-way operation. Once you `eject`, you can’t go back!**

If you aren’t satisfied with the build tool and configuration choices, you can `eject` at any time. This command will remove the single build dependency from your project.

Instead, it will copy all the configuration files and the transitive dependencies (Webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you’re on your own.

You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.

## Learn More

You can learn more in the [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started).

To learn React, check out the [React documentation](https://reactjs.org/).






CHANGES.md
## Changes in version 0.3.2

### Enhancements

* It is now possible to change the base map layer from settings. (#30)
* Different deployments can now be configured through `.env.local` file. (#24)

### Other changes

* Updated code to use latest React 16.8 features i.e. `React.FC` with hooks (#111)

* Users can now login. (#22)

## Changes in version 0.3.1

Fixed translation of legal agreement.

## Changes in version 0.3.0

### Enhancements

* It is now possible to adjust the color bar min/max value and the colors. (#29)

### Other changes

* Now requires xcube 0.3.0 as backend.

## Changes in version 0.2.1

### Enhancements

* A new setting allows to turn image smoothing of image layers on and off.
  By default it is off now, so that crisp image pixels are shown without any blurring. (#86)
* The color bar legend uses 5 tick marks instead of the two.
* Added simple scale indicator to map.

### Other changes

* Now using OpenLayers 6.1.1.


## Changes in version 0.2.0

* Users can now draw polygons and circles and show respective time-series. (#32)
  In total there are now 4 new interactions with the map:
  - Select a geometry
  - Draw a point
  - Draw a geometry
  - Draw a circle
* Improved selection of places:
  - If a place is selected in the list it is zoomed to and highlighted in the map;
  - Places can now be selected in the map if the "Select" map interaction is active;
  - Places are also selected if a time-series line is clicked.
    (However, that doesn't work nicely yet due to issues in the Recharts lib).
* Minor fixes:
  * Fixed (actually avoided) problem indicated by text "something went wrong" appearing instead of map
    after server change. 
  * Changed user setting "show graph after adding point" to be on by default.
  * Added translations for message "server did not respond".
  * Time-series charts now have constant spacing;
  * A loading icon is shown, while time-series are being loaded.
  * Restrict zooming out, so that only a single world is shown. (#79)
  * Added legal agreement to inform about using HTML local storage. (#77)
  * Colors used for geometries in the map and for lines in the time-series charts are now exactly the same.
  * Time-series charts now have constant spacing
  * Restrict zooming out, so that only a single world is shown. (#79)
  * Added legal agreement to inform about using HTML local storage. (#77)
  * Corrected Romanian translations
* Users can now generate time series for any selected variable and any selected place. (#50)
  User places can now be also removed, which will remove related time series too.
  Removing time series, on the other hand, does no longer remove user added places.
* Selection of exact time steps has been improved. Introduced new time picker and user can also
  select move forward and backward in time by single steps. (#58)
* User can now browse settings and adjust them according to preferences in a new settings dialog:
  * xcube servers;
  * UI language;
  * whether to connect points by line (#62);
  * whether to show error bars / whether to compute stdev yes/no;
  * auto-play speed / time interval in millis;
  * whether to auto-add time-series chart if point is added (#46);
  * system-information (#93).
* Removed dummy app bar icons ("Notifications", "Avatar") and menu entry ("Settings...")
* Place groups dropdown menus are no longer displayed if a dataset has no 
  place groups and the places dropdown is no longer shown if no place group is selected. 
* Place groups (GeoJSON feature collections) are now only loaded if selected which significantly
  increases viewer loading time for server configurations whose data cubes
  are associated with lots of vector data. (#61)
* Viewer loads now in the MS Edge browser (#59)
* Fixed problem where viewer did not showInMap to selected place if that is a point location. (#52)
* Time-series are now loaded in increments so user see constantly growing time-series graph
  instead of doing nothing for tens of seconds until the server returns the complete series (#38)
* When animating through time, the index into a dataset's time coordinates array is incremented
  rather than incrementing current UI time by constant delta (#40)
* The viewer now allows for multiple time-series groups based on the variable's units (#27)
* Fixed a problem where selecting data points in time-series didn't show any data (#41)
* Added translation for Italiano and Rumeno (#25)
* Time-series are now plotted with standard deviation, if available e.g. for weekly means (#19)
* Colors of of time-series line charts are now synchronised with fill colors of points in map (#17)
* Map points and extracted time-series are now synchronised (#16):
  - When in "multi" mode, if all time-series are removed, also remove all related points in map
  - When not in "multi" mode, delete old point in map (if any) once new point is added
* It is now possible to switch between different servers (#14)

## Changes in version 0.1

Initial development version.

service-worker.js
"use strict";var precacheConfig=[["./index.html","8c081bb567e7d87435988c1a6878e0f0"],["./static/css/main.74041a7f.css","bb7570a5fd4159c0e82d976d45b7d7b1"],["./static/js/main.52e61a8a.js","878e80de5f3f7958e195ab7e49d4111c"],["./static/media/logo.9f22c0b4.png","9f22c0b4ad07cefbcfc7ca6978e986be"]],cacheName="sw-precache-v3-sw-precache-webpack-plugin-"+(self.registration?self.registration.scope:""),ignoreUrlParametersMatching=[/^utm_/],addDirectoryIndex=function(e,t){var n=new URL(e);return"/"===n.pathname.slice(-1)&&(n.pathname+=t),n.toString()},cleanResponse=function(t){return t.redirected?("body"in t?Promise.resolve(t.body):t.blob()).then(function(e){return new Response(e,{headers:t.headers,status:t.status,statusText:t.statusText})}):Promise.resolve(t)},createCacheKey=function(e,t,n,r){var a=new URL(e);return r&&a.pathname.match(r)||(a.search+=(a.search?"&":"")+encodeURIComponent(t)+"="+encodeURIComponent(n)),a.toString()},isPathWhitelisted=function(e,t){if(0===e.length)return!0;var n=new URL(t).pathname;return e.some(function(e){return n.match(e)})},stripIgnoredUrlParameters=function(e,n){var t=new URL(e);return t.hash="",t.search=t.search.slice(1).split("&").map(function(e){return e.split("=")}).filter(function(t){return n.every(function(e){return!e.test(t[0])})}).map(function(e){return e.join("=")}).join("&"),t.toString()},hashParamName="_sw-precache",urlsToCacheKeys=new Map(precacheConfig.map(function(e){var t=e[0],n=e[1],r=new URL(t,self.location),a=createCacheKey(r,hashParamName,n,/\.\w{8}\./);return[r.toString(),a]}));function setOfCachedUrls(e){return e.keys().then(function(e){return e.map(function(e){return e.url})}).then(function(e){return new Set(e)})}self.addEventListener("install",function(e){e.waitUntil(caches.open(cacheName).then(function(r){return setOfCachedUrls(r).then(function(n){return Promise.all(Array.from(urlsToCacheKeys.values()).map(function(t){if(!n.has(t)){var e=new Request(t,{credentials:"same-origin"});return fetch(e).then(function(e){if(!e.ok)throw new Error("Request for "+t+" returned a response with status "+e.status);return cleanResponse(e).then(function(e){return r.put(t,e)})})}}))})}).then(function(){return self.skipWaiting()}))}),self.addEventListener("activate",function(e){var n=new Set(urlsToCacheKeys.values());e.waitUntil(caches.open(cacheName).then(function(t){return t.keys().then(function(e){return Promise.all(e.map(function(e){if(!n.has(e.url))return t.delete(e)}))})}).then(function(){return self.clients.claim()}))}),self.addEventListener("fetch",function(t){if("GET"===t.request.method){var e,n=stripIgnoredUrlParameters(t.request.url,ignoreUrlParametersMatching),r="index.html";(e=urlsToCacheKeys.has(n))||(n=addDirectoryIndex(n,r),e=urlsToCacheKeys.has(n));var a="./index.html";!e&&"navigate"===t.request.mode&&isPathWhitelisted(["^(?!\\/__).*"],t.request.url)&&(n=new URL(a,self.location).toString(),e=urlsToCacheKeys.has(n)),e&&t.respondWith(caches.open(cacheName).then(function(e){return e.match(urlsToCacheKeys.get(n)).then(function(e){if(e)return e;throw Error("The cached response that was expected is missing.")})}).catch(function(e){return console.warn('Couldn\'t serve response for "%s" from cache: %O',t.request.url,e),fetch(t.request)}))}});

src/config.ts
import {
    blue,
    green,
    grey,
    purple,
    red,
    pink,
    yellow,
    orange,
    cyan,
    indigo,
    deepPurple
} from '@material-ui/core/colors';
import { Color, PaletteType } from '@material-ui/core';
import { PaletteColorOptions } from '@material-ui/core/styles/createPalette';

import { LanguageDictionary } from './util/lang';
import { getQueryParameterByName } from './util/qparam';
import lang from './resources/lang.json';

const version = '0.5.0-dev.0';

const defaultApiServerUrl = process.env.REACT_APP_XCUBE_API;

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

interface Branding {
    appBarTitle: string;
    windowTitle: string;
    defaultApiServerId: string;
    defaultApiServerName: string;
    defaultApiServerUrl: string;
    themeName: 'dark' | 'light';
    primaryColor: PaletteColorOptions;
    secondaryColor: PaletteColorOptions;
    headerBackgroundColor?: string;
    logoPath: any;
    logoWidth: number;
    baseMapUrl?: string;
    defaultAgg?: 'median' | 'mean';
}

const brandings: { [name: string]: Branding } = {
    'default': {
        appBarTitle: 'xcube Viewer',
        windowTitle: 'xcube Viewer',
        defaultApiServerId: 'local',
        defaultApiServerName: 'Local Server',
        defaultApiServerUrl: defaultApiServerUrl || 'http://localhost:8080',
        themeName: 'dark',
        primaryColor: blue,
        secondaryColor: pink,
        headerBackgroundColor: undefined,
        logoPath: require('./resources/default/logo.png'),
        logoWidth: 32,
        baseMapUrl: 'http://services.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
        defaultAgg: 'mean',
    },
    'esdl': {
        appBarTitle: 'ESDL Viewer',
        windowTitle: 'ESDL Viewer',
        defaultApiServerId: 'esdl',
        defaultApiServerName: 'ESDL Server',
        defaultApiServerUrl: 'https://xcube.earthsystemdatalab.net',
        themeName: 'light',
        primaryColor: grey,
        secondaryColor: pink,
        headerBackgroundColor: grey['50'],
        logoPath: require('./resources/esdl/logo.png'),
        logoWidth: 64,
        baseMapUrl: 'http://services.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
        defaultAgg: 'mean',
    },
    'eodatabee': {
        appBarTitle: 'Demo Viewer',
        windowTitle: 'EODataBee Demo Viewer',
        defaultApiServerId: 'eodatabee',
        defaultApiServerName: 'EODataBee Server',
        defaultApiServerUrl: defaultApiServerUrl || 'https://xcube2.dcs4cop.eu/dcs4cop-dev/api/latest',
        themeName: 'dark',
        primaryColor: {
            light: '#ffd77c',
            main: '#ffc942',
            dark: '#b78c2f',
            contrastText: '#fff',
        },
        secondaryColor: pink,
        headerBackgroundColor: undefined,
        logoPath: require('./resources/eodatabee/logo.png'),
        logoWidth: 150,
        baseMapUrl: 'http://a.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png',
        defaultAgg: 'mean',
    },
    'cyanoalert': {
        appBarTitle: '',
        windowTitle: 'CyanoAlert Viewer',
        defaultApiServerId: 'cyanoalert',
        defaultApiServerName: 'CyanoAlert Server',
        defaultApiServerUrl: defaultApiServerUrl || 'https://cyanoalert.brockmann-consult.com/api/latest',
        themeName: 'dark',
        primaryColor: {
            light: '#ceef64',
            main: '#9abc31',
            dark: '#688c00',
            contrastText: '#fff',
        },
        secondaryColor: deepPurple,
        headerBackgroundColor: undefined,
        logoPath: require('./resources/cyanoalert/logo.png'),
        logoWidth: 120,
        baseMapUrl: 'https://tiles.wmflabs.org/bw-mapnik/{z}/{x}/{y}.png',
        defaultAgg: 'median',
    },
};

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

const brandingName = process.env.REACT_APP_BRANDING || 'default';
const branding = brandings[brandingName];

if (!branding) {
    throw new Error(`illegal branding "${brandingName}"!`);
}

if (branding.windowTitle) {
    document.title = branding.windowTitle;
}

export const I18N = new LanguageDictionary(lang);


interface TileAccess {
    param: string;
    token: string;
}


const defaultApiServer = {
    id: branding.defaultApiServerId,
    name: getQueryParameterByName(null, 'serverName', branding.defaultApiServerName)!,
    url: getQueryParameterByName(null, 'serverUrl', branding.defaultApiServerUrl)!,
};

const apiServers = [
    {...defaultApiServer},
];


export function getVersion() {
    return version;
}

export function getBrandingName() {
    return brandingName;
}

export function getBranding() {
    return branding;
}

export function getDefaultApiServer() {
    return defaultApiServer;
}

export function getApiServers() {
    return apiServers;
}

const userPlaceColors: { [name: string]: Color } = {
    blue,
    green,
    grey,
    purple,
    red,
    pink,
    yellow,
    orange,
    cyan,
    indigo,
    deepPurple,
};
export const userPlaceColorNames = Object.keys(userPlaceColors);


export function getStrokeShade(paletteType: PaletteType) {
    return paletteType === 'light' ? 800 : 400;
}

export function getUserPlaceColorName(index: number): string {
    return userPlaceColorNames[index % userPlaceColorNames.length];
}

export function getUserPlaceColor(colorName: string, paletteType: PaletteType): string {
    const shade = getStrokeShade(paletteType);
    return userPlaceColors[colorName][shade];
}

// See resources/maps.json
const tileAccess: { [name: string]: TileAccess } = {
    'Mapbox': {
        param: 'access_token',
        token: 'pk.eyJ1IjoiZm9ybWFuIiwiYSI6ImNrM2JranV0bDBtenczb2szZG84djh6bWUifQ.q0UKwf4CWt5fcQwIDwF8Bg'
    },
};


export function getTileAccess(groupName: string) {
    return tileAccess[groupName];
}